Packet++/header/SomeIpLayer.h

```diff
@@ -0,0 +1,341 @@
+
+	/**
+	 * Set the Session ID
+	 * @param[in] sessionID sessionID to set
+	 */
+	void setSessionID(uint16_t sessionID);
+
+	/**
+	 * Get the clientID
+	 * @return uint16_t returned in host endian
+	 */
+	uint16_t getClientID() const;
+
+	/**
+	 * Set the Client ID
+	 * @param[in] clientID clientID to set
+	 */
+	void setClientID(uint16_t clientID);
+
+	/**
+	 * Get the protocolVersion
+	 * @return uint8_t
+	 */
+	uint8_t getProtocolVersion() const;
+
+	/**
+	 * Set the Protocol Version
+	 * @param[in] version version to set
+	 */
+	void setProtocolVersion(uint8_t version);
+
+	/**
+	 * Get the interfaceVersion
+	 * @return uint8_t
+	 */
+	uint8_t getInterfaceVersion() const;
+
+	/**
+	 * Set the Interface Version
+	 * @param[in] version version to set
+	 */
+	void setInterfaceVersion(uint8_t version);
+
+	/**
+	 * Get the message type
+	 * @return uint8_t
+	 */
+	uint8_t getMessageType() const;
+
+	/**
+	 * Set the Message Type
+	 * @param[in] type Type to set
+	 */
+	void setMessageType(MsgType type);
+
+	/**
+	 * Set the Message Type
+	 * @param[in] type Type to set
+	 */
+	void setMessageType(uint8_t type);
+
+	/**
+	 * Get the returnCode
+	 * @return uint8_t
+	 */
+	uint8_t getReturnCode() const;
+
+	/**
+	 * Set the returnCode
+	 * @param[in] returnCode ReturnCode to set
+	 */
+	void setReturnCode(uint8_t returnCode);
+
+	/**
+	 * Set the length field of the SOME/IP header
+	 * @param[in] payloadLength Length of the payload
+	 */
+	void setPayloadLength(uint32_t payloadLength);
+
+	/**
+	 * Get the Header Length With Payload, as is stated in the length field of the SOME/IP header.
+	 * @return uint32_t Length of the complete SOME/IP Pdu, including header and payload
+	 */
+	uint32_t getHeaderLengthWithPayload() const;
+
+	/**
+	 * @return A pointer for the layer payload, meaning the first byte after the header
+	 */
+	uint8_t *getPduPayload() const { return m_Data + sizeof(someiphdr);	}
+
+	/**
+	 * @return The size in bytes of the payload
+	 */
+	size_t getPduPayloadSize() const { return getHeaderLengthWithPayload() - sizeof(someiphdr); }
+
+	/**
+	 * Get the Length of the SOME/IP header inc payload
+	 * @return size_t
+	 */
+	size_t getHeaderLen() const;
+
+	/**
+	 * Sets missing fields and calculates length
+	 */
+	void computeCalculateFields();
+
+	/**
+	 * Identifies next layer and tries to create it
+	 */
+	void parseNextLayer();
+
+	/**
+	 * @return The string representation of the SOME/IP layer
+	 */
+	std::string toString() const;
+
+	/**
+	 * @return The OSI model layer of this layer
+	 */
+	OsiModelLayer getOsiModelLayer() const { return OsiModelApplicationLayer; }
+
+private:
+	static const uint8_t SOMEIP_PROTOCOL_VERSION = 1;
+
+	/* Using unordered_set since insertion and search should be almost constant time */
+	static std::unordered_set<uint16_t> m_SomeIpPorts;
+
+	bool subProtocolFollows() const;
+	static void splitUint32Id(uint32_t uint32Id, uint16_t &uint16IdUpper, uint16_t &uint16IdLower);

```

Comment:
I just noticed this is a helper function that is only used within `SomeIpLayer.cpp`, hence we don't need to keep it as a private class method, instead it can be a method inside `SomeIpLayer.cpp` (without a class)